### ISSUE TITLE
Change Request: allow unknown properties in `languageOptions`

### ISSUE THREAD

**Issue by fasttime** [eslint/eslint#19615]
### ESLint version

v9.24.0

### What problem do you want to solve?

Currently, the `js` language rejects unknown properties on the `languageOptions` object with an error, e.g.:

```js
import { ESLint } from "eslint";

const eslint = new ESLint({ baseConfig: { languageOptions: { foo: 42 } } });
await eslint.lintFiles("*.js");
```

throws

    TypeError: Key "languageOptions": Unexpected key "foo" found.

This behavior contrasts with  `@eslint/json`, `@eslint/css` and `@eslint/markdown` which all allow unknown properties on the  `languageOptions` object. By extending the `LanguageOptions` core type, the `Linter.LanguageOptions` type does not forbid adding unknown properties:

```ts
// test.ts
import type { Linter } from "eslint";

const languageOptions: Linter.LanguageOptions = { 42: "foo" };
```

### What do you think is the correct solution?

Allowing unknown properties in the `languageOptions` object for the `js` language for consistency and type correctness.

### Participation

- [x] I am willing to submit a pull request for this change.

### Additional comments

refs https://github.com/eslint/markdown/pull/328/files#r2026262495

**Comment by nzakas** [comment]
Seems like this wouldn't be a breaking change so I'm :+1:.

@mdjermanovic ?

**Comment by mdjermanovic** [comment]
> This behavior contrasts with `@eslint/json`, `@eslint/css` and `@eslint/markdown` which all allow unknown properties on the `languageOptions` object.

Allowing unknown properties was necessary for other languages at the time because ESLint always automatically applied the default JS language options. Now that ESLint supports `Language#defaultOptions`, and the JS language default options have been extracted into `JSLanguage.defaultOptions` (https://github.com/eslint/eslint/commit/381c32b6975fa3208c62ca2b1052eb87182ed731) so they apply only when the `js` language is used, allowing unknown language options might not be necessary. I'd rather keep rejecting unknown properties in the `js` language, and maybe consider doing the same in other languages, because specifying an unknown option is likely an error in the config (e.g., a typo) for which throwing a config error seems better than silently ignoring the option.

**Comment by nzakas** [comment]
I think there's still a significant risk of unknown `languageOptions` leaking into other languages. There are still a lot of examples on the web that show setting up a config without using `files`. Considering the number of issues we've had open because someone accidentally enabled a JS rule on a Markdown or JSON file, we'd end up with the same problem for `languageOptions` if we did strict validation in plugins.

**Comment by nzakas** [comment]
> Sounds reasonable to me. But I'd still keep the strict validation in the `js` language as it seems unlikely that language options or any other config for other languages would be specified without `files`.

That's a fair point. I think we can keep the strict validation in the `js` language. @fasttime?

**Comment by fasttime** [comment]
Thanks for the feedback. The reason for this proposal was to make the handling of `languageOptions` more consistent across different languages, as I didn't actually consider the consequences of this change for users.

> I'd rather keep rejecting unknown properties in the `js` language, and maybe consider doing the same in other languages, because specifying an unknown option is likely an error in the config (e.g., a typo) for which throwing a config error seems better than silently ignoring the option.

I'm wondering at this point, if unknown language options are likely an error (to the extent that we'd like to forbid them for the `js` language) if it wouldn't make sense to at least discourage their usage for other languages. This could be done for example with a runtime warning or a note in the documentation.

**Comment by nzakas** [comment]
I don't think we should do anything more aggressive with unknown `languageOptions`. Because these can be merged from multiple configs, it's very easy to inadvertently introduce an option that is meant for another language. Further, there might not be an easy way to remove the extra option.

I still think the best course for language plugins is just to ignore any options it doesn't recognize to reduce unnecessary burden on the user.
